pallets/pips/src/lib.rs

```diff
@@ -1341,23 +1344,23 @@ impl<T: Config> Module<T> {
         id: PipId,
         new_state: ProposalState,
     ) -> ProposalState {
-        <Proposals<T>>::mutate(id, |proposal| {
-            if let Some(ref mut proposal) = proposal {
+        <ProposalStates>::mutate(id, |proposal_state| {
+            if let Some(ref mut proposal_state) = proposal_state {
                 // Decrement active count, if the `new_state` is not active.
                 if !Self::is_active(new_state) {
-                    Self::decrement_count_if_active(proposal.state);
+                    Self::decrement_count_if_active(*proposal_state);
                 }
-                proposal.state = new_state;
             }
+            *proposal_state = Some(new_state);
```

Comment:
This should also be inside the `if` statement (line 1354 -> 1350).  The reason is that the state should not be changed if the old state was `None`.  `update_proposal_state` should only be called on a pip that already has a state.
`*proposal_state = Some(new_state);`